### PR TITLE
[Scala3] Redux: Convert cloneType to an extension method

### PR DESCRIPTION
**Pull request by adkian-sifive**
Redux of #3771. Turns out we need this for Scala3 after all

Ran into errors related to this while adding Scala3 cross-compilation support. Here's a sample of what an error with Dotty looks like:

```
[error] -- [E007] Type Mismatch Error: chisel3/core/src/main/scala/chisel3/AggregateImpl.scala:838:4 
[error] 838 |    clone
[error]     |    ^^^^^
[error]     |    Found:    (clone : chisel3.Record & chisel3.RecordImpl)
[error]     |    Required: (RecordImpl.this : chisel3.Record & chisel3.RecordImpl)
```

I suspect this is due to the tightening of the type system in Scala3, specifically changes which make `this.type` a more stricter representation of the current instance. I think this SIP might have a more detailed specification of the related changes: https://docs.scala-lang.org/sips/42.type.html#related-scala-issues-resolved-by-the-literal-types-implementation

### Contributor Checklist

- [ ] Did you add Scaladoc to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [ ] Did you delete any extraneous printlns/debugging code?
- [ ] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [ ] Did you request a desired merge strategy?
- [ ] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- API Modification
- Internal or build-related (includes code refactoring/cleanup)

#### Desired Merge Strategy

- Squash: The PR will be squashed and merged (choose this if you have no preference).

#### Release Notes
Move cloneType to an extension method in preparation for Scala3 

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels? (Select the most appropriate one based on the "Type of Improvement")
- [ ] Did you mark the proper milestone (Bug fix: `3.6.x`, `5.x`, or `6.x` depending on impact, API modification or big change: `7.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

**Comment by jackkoenig** [comment]
> Turns out we need this for Scala3 after all

Can you explain more about _why_ we need this? I know we saw errors related to it, but is there documentation or even a thread somewhere in Scala 3 development explaining the change?

**Comment by adkian-sifive** [comment]
@jackkoenig updated the comment above  with some more info

File: core/src/main/scala/chisel3/BitsImpl.scala [hunk 1]
```diff
@@ -558,7 +556,7 @@ private[chisel3] trait AsyncResetImpl extends Element { self: AsyncReset =>
 
   override def toString: String = stringAccessor("AsyncReset")
 
-  def cloneType: this.type = AsyncReset().asInstanceOf[this.type]
+  override def _cloneType: Data = AsyncReset()
```
Comment (mwachs5) [inline core/src/main/scala/chisel3/BitsImpl.scala:559]:
```suggestion
  override def _cloneType: AsyncReset = AsyncReset()
```

? perhaps? Why is this one different than the others

File: core/src/main/scala/chisel3/ClockImpl.scala [hunk 2]
```diff
@@ -13,7 +13,7 @@ private[chisel3] trait ClockImpl extends Element {
 
   override def toString: String = stringAccessor("Clock")
 
-  def cloneType: this.type = Clock().asInstanceOf[this.type]
+  override def _cloneType: Data = Clock()
```
Comment (mwachs5) [inline core/src/main/scala/chisel3/ClockImpl.scala:16]: why isn't this one `Clock`?

File: core/src/main/scala/chisel3/DataImpl.scala [hunk 3]
```diff
@@ -1059,6 +1044,33 @@ private[chisel3] trait ObjectDataImpl {
       }
     }
   }
+
+  implicit class DataExtensions[T <: Data](self: T) {
+
+    /** Internal API; Chisel users should look at chisel3.chiselTypeOf(...).
```
Comment (mwachs5) [inline core/src/main/scala/chisel3/DataImpl.scala:1050]: If this is an internal API, can we make it private to chisel?

**Review by mwachs5 — COMMENTED** [review]
high level, why are we sometimes having the `_cloneType` return type be Data and other times we are having the explicit type?

**Comment by jackkoenig** [comment]
I'm pretty sure `this.type` is fine in Scala 3, there's just a problem we run into occasionally which I've captured in a scala-cli script:
```scala
//> using scala "3.3.4"

trait Data {
  var name: String = "<default>"

  def cloneType: this.type

  override def toString: String = s"Data($name)"
}

class UInt(_name: String) extends Data {
  this.name = _name

  override def cloneType: this.type =
    new UInt(s"Clone of ${_name}").asInstanceOf[this.type]

  def withName(_name: String): this.type = {
    // Without `: this.type` we get this compile error below
    // Found:    (clone : UInt)
    // Required: (UInt.this : UInt)
    //     clone
    //     ^^^^^
    val clone: this.type = this.cloneType
    clone.name = _name
    clone
  }
}

object Main extends App {
  val u = new UInt("u")
  val v = u.withName("v")
  println(u)
  println(v)
}
```

It just requires more type annotations when using `this.type` to return a `this.type`. This should be fine for Scala 3 in Chisel.

**Pull request closed without merging**